docs/spec.md

```diff
@@ -0,0 +1,406 @@
+2. **Retrieve**: Get files from Storacha by their Content ID (CID) through the Storacha HTTP trustless Gateway
+3. **Authentication & Authorization**: Optional authentication via bearer tokens
+
+#### Upload Operation
+
+- **Multiple File Support**: The upload operation can handle one or more files in a single request
+- **Universal File Types**: Supports any file type, which is automatically converted to a blob by the Storacha client
+- **Content Addressing**: Returns Content IDs (CIDs) for uploaded files, enabling permanent, verifiable references
+- **Space Delegation**: Files are uploaded to the space authorized by the provided delegation in the server configuration
+- **Flexible Space Delegation**: Can accept delegation proof in request headers or body, with runtime delegations taking priority over server configuration
+
+#### Retrieve Operation
+
+- **CID-Based Access**: Files are retrieved using their unique Content IDs (CIDs)
+- **Gateway Access**: Retrieval happens through the Storacha HTTP trustless gateway (https://storacha.link) rather than the client instance, but the users can provide a custom gateway in the server configuration if needed
+- **Format Conversion**: Files are returned in a format compatible with the MCP specification (base64 data URIs)
+- **Metadata Preservation**: Content types and other metadata are preserved and returned with the file
+- **Public Access**: No authentication required for retrieval, as CIDs are self-authenticating
+
+#### Authentication & Authorization
+
+- **Bearer Token Authentication**: Optional API key authentication using bearer tokens to control access to the upload operation
+- **Delegation Priority**: Runtime delegations provided in requests take precedence over server pre defined delegation configuration
+- **Flexible Authentication**: Can operate with or without authentication depending on use case and server configuration
+
+#### MCP Compliance
+
+- **Discovery Endpoint**: Implements the `.well-known/mcp.json` endpoint for client discovery
+- **Standard Tool Format**: Operations are exposed as MCP tools with standard parameter schemas
+- **Schema Validation**: All parameters are validated using Zod schemas
+- **Consistent Error Handling**: Standardized error responses following the MCP format
+
+## Requirements
+
+### Server Configuration
+
+The Storacha MCP server requires the following configuration:
+
+- **SHARED_ACCESS_TOKEN**: Optional access token used to authenticate upload operations (if not set, authentication is disabled)
+- **PRIVATE_KEY**: Agent private key for w3up-client (for default delegation)
+- **DELEGATION**: Default delegation authorized to store content
+- **PORT**: Port to run the server on (default: 3000)
+- **HOST**: Host to bind to (default: '0.0.0.0')
+- **GATEWAY_URL**: Custom gateway URL (default: 'https://storacha.link')
+- **MAX_FILE_SIZE_MB**: Maximum file size in MB (default: 100MB)
+  - TODO: double check this limitation in the Storacha client
```

Comment:
What is this limitation? There are no restrictions in the Storacha client AFAIK.

Reply:
I found a code snippet in the w3infra/upload service about the max file size, it was something around 5GB, not sure if that is enforced or not, so I need to double check that with you.

Reply:
As we discussed, there is no limitation on the file size. This is the max size of the shard, we use 128MB shard size in the client, and it can be set up to 5GB, but that doesn't block us from uploading any file size.